src/couch_replicator/src/couch_replicator_ids.erl

```diff
@@ -125,3 +133,154 @@ get_rep_endpoint(_UserCtx, #httpdb{url=Url, headers=Headers, oauth=OAuth}) ->
+get_v4_url_info(Url) ->
+    case ibrowse_lib:parse_url(Url) of
+        {error, invalid_uri} ->
+            % Tolerate errors here to avoid a bad user document
+            % crashing the replicator
+            {undefined, Url, undefined, undefined};
+        #url{
+            protocol = Schema,
+            username = User,
+            host = Host,
+            port = Port,
+            path = Path
+        } ->
+            NonDefaultPort = get_non_default_port(Schema, Port),
+            {User, Host, NonDefaultPort, Path}
+    end.
+
+
+% Use 0 if a port is default to keep value an integer
+get_non_default_port(https, 443) ->
+    0;
+get_non_default_port(http, 80) ->
+    0;
+get_non_default_port(_Schema, Port) ->
+    Port.
+
+
+-ifdef(TEST).
+
+-include_lib("eunit/include/eunit.hrl").
+
+http_v4_endpoint_test_() ->
+    [?_assertMatch({remote, User, Host, Port, Path, HeadersNoAuth, undefined},
+        get_v4_endpoint(nil, #httpdb{url = Url, headers = Headers})) ||
+            {{User, Host, Port, Path, HeadersNoAuth}, {Url, Headers}} <- [
+                {
+                    {undefined, "host", 0, "/", []},
+                    {"http://host", []}
+                },
+                {
+                    {undefined, "host", 0, "/", []},
+                    {"https://host", []}
+                },
+                {
+                    {undefined, "host", 1, "/", []},
+                    {"http://host:1", []}
+                },
+                {
+                    {undefined, "host", 2, "/", []},
+                    {"https://host:2", []}
+                },
+                {
+                    {undefined, "host", 0, "/", [{"h","v"}]},
+                    {"http://host", [{"h","v"}]}
+                },
+                {
+                    {undefined, "host", 0, "/a/b", []},
+                    {"http://host/a/b", []}
+                },
+                {
+                    {"user", "host", 0, "/", []},
+                    {"http://user:pass@host", []}
+                },
+                {
+                    {"user", "host", 0, "/", []},
+                    {"http://user:newpass@host", []}
+                },
+                {
+                    {"user", "host", 0, "/", []},
+                    {"http://host", [basic_auth("user","pass")]}
+                },
+                {
+                    {"user", "host", 0, "/", []},
+                    {"http://host", [basic_auth("user","newpass")]}
+                },
+                {
+                    {"user1", "host", 0, "/", []},
+                    {"http://user1:pass1@host", [basic_auth("user2","pass2")]}
+                },
+                {
+                    {"user", "host", 0, "/", [{"h", "v"}]},
+                    {"http://host", [{"h", "v"}, basic_auth("user","pass")]}
+                }
+
+        ]
+    ].
+
+
+basic_auth(User, Pass) ->
+    B64Auth = ibrowse_lib:encode_base64(User ++ ":" ++ Pass),
```

Comment:
Good point. `base64:encode` produces binaries, but there is` base64:encode_to_string` I'll use that